sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+```javascript
+// Angular Component file example
+/* setup:
+ 0. Identify the component in the DOM;
+  inspect and find the corresponding <app.. tag,
+  then find the component in src
+ 1. Insert conditional */
+constructor(
+  /* ... */
+) {
+  /* if running inside Cypress tests, set the component
+  may need // @ts-ignore initially */
+  if (window.Cypress) {
+    window.yourComponent = this;
+  }
+}
+
+// at ../../support/app-actions.ts helper file:
+
+/** yields  window.yourComponent */
+export const yourComponent = () =>
+  cy.window().should('have.property', 'yourComponent');
+
+/** yields the data property on your component */
+export const getSomeListData = () =>
+ yourComponent().should('have.property', 'data');
+```
+After this at DevTools see what that component allows for properties, or in the component code see what functions you can .invoke()
+
+Check out [the presentation slide](https://cypress.slides.com/cypress-io/siemens-case-study#/12/3/4) for a code sample utilizing app actions with visual testing.
+
+#### Another example on app actions utilizing states, using [Building Operator](https://new.siemens.com/us/en/products/buildingtechnologies/automation/talon/software/building-operator.html?stc=ussi100451&sp_source=ussi100451&&s_kwcid=AL!464!3!435315652461!b!!g!!%2Bbuilding%20%2Boperator&ef_id=CjwKCAjw8df2BRA3EiwAvfZWaAsQmgot5Ph-nGBB8rW1QLLr870q2HW-qzMKhqtQb1QvlPBVJxho5BoCmtMQAvD_BwE:G:s) building control product of Siemens.
+
+In the below state diagram there are 3 states. We begin where both left and right panes exist. If the right pane is deleted (delete point / red flow), only the left pane exists. If the left pane is deleted (delete device - blue flow), both panes go away and the UI get redirected.
+
+![deleting building points and controllers](../../assets/images/ui-state/delete-states.PNG)
+
+Testing the UI, you might choose to delete the right pane (red flow) and then in another test you might choose to delete the left pane (blue flow). This leaves out 1 final path through the state diagram where right pane and then left pane are deleted one at a time.
+
+We already covered deleting the right pane in a UI test (red path). Why not avoid repeating this test and utilize app app actions, getting access to the delete function in the source code and using `cy.invoke()` to call it?
+
+```javascript
+it('Component test: delete right pane and then left', () => {
+  /* tests a SEQUENCE not covered with UI tests
+   * tests a COMBINATION of components */
+  appAction.deleteRightPane();
+  cy.window().should('not.have.property', 'rightPaneComponent');
+  cy.window().should('have.property', 'leftPaneComponent');
+
+  appAction.deleteLeftPane();
+  cy.window().should('not.have.property', 'leftPaneComponent');
+  cy.window().should('not.have.property', 'rightPaneComponent');
+  cy.url().should('match', redirectRoute);
+});
+```
+<br/><br/>
```

Comment:
Im my opinion, the pane deletion is not the best example because:
- it does not show the original "pane deletion" test, is it made by 20 UI steps? Or just 2?
- it's an uncommon case so the reader could not feel the pain of reaching the UI state manually every time (a form filling process could give the reader a better idea)

Because of that, I think that an example with a form could give a better idea, for example: testing the post removal in [Conduit](https://angular.realworld.io/) needs the user to be logged and the post to be created. The `appActions.login()` and `appActions.createPost()` could be easier to be understood, what do you think?

Reply:
Maybe you are right, but this is a real life example.

I cannot help with an example I have not implemented.

If you have one, feel free to substitute it instead. Or you could add it on top of this? The more the better.

Reply:
I tried to shorten it as much as possible, take a look at my new comment

Reply:
looks good